### PR TITLE
fix: database corruption when stopping node with cache enabled

### PR DESCRIPTION
**Pull request by luislhl**
fixes: https://github.com/HathorNetwork/hathor-core/issues/487

### Acceptance Criteria
We shouldn't corrupt the database when stopping the node.

There is a better description of the error and the exact scenario that triggered it in the issue above.

I've only observed this when running a local fullnode and stopping it with `Ctrl + C`, but I'm not sure if this couldn't also happen when running it with Docker in the cloud.

### CODE REVIEW

**Review by jansegre — APPROVED** [review]

File: hathor/manager.py [hunk 1]
```diff
@@ -318,6 +318,10 @@ def stop(self) -> Deferred:
             wait_stratum = self.stratum_factory.stop()
             if wait_stratum:
                 waits.append(wait_stratum)
+
+        if isinstance(self.tx_storage, TransactionCacheStorage):
```
Comment (msbrogli) [inline hathor/manager.py:322]: Why don't you add the `flush()` method to all storages? All others would do nothing while the `TransactionCacheStorage` would.

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/manager.py (hunk 1 above)
Comment (luislhl) [reply hathor/manager.py:322]: Makes sense, will do it.

**Review by msbrogli — APPROVED** [review]

File: hathor/transaction/storage/transaction_storage.py [hunk 2]
```diff
@@ -1116,3 +1123,10 @@ def get_blocks_before(self, hash_bytes: bytes, num_blocks: int = 100) -> List[Bl
                     used.add(parent_hash)
                     pending_visits.append(parent_hash)
         return result
+
+    def flush(self) -> None:
+        """Flushes the storage. It's called during shutdown of the node, for instance.
+
+           Should be implemented by storages that provide some kind of in-memory cache
+        """
```
Comment (jansegre) [inline hathor/transaction/storage/transaction_storage.py:1131]: I think we can remove the docstring here, so it's not duplicated from the base interface method.

File: hathor/transaction/storage/transaction_storage.py [hunk 3]
```diff
@@ -749,6 +749,13 @@ def update_last_started_at(self, timestamp: int) -> None:
                 continue
             self.set_index_last_started_at(index_db_name, timestamp)
 
+    def flush(self) -> None:
```
Comment (jansegre) [inline hathor/transaction/storage/transaction_storage.py:752]: This should probably be annotated with `@abstractmethod` so we don't forget to implement it.

**Review by jansegre — APPROVED** [review]
I think it looks good, all changes I suggested are really minor.

**Pull request merged**